### ISSUE TITLE
[GSPMD] partial `shard_map` with `out_specs` referencing a manual axis fails to lower since bbd2b2461

### ISSUE THREAD

**Issue by hugues-ant** [jax-ml/jax#38845]
## Description

Since bbd2b2461929a83144d1c3719bd607b8c9cd28eb (first appears in nightly `0.11.0.dev20260627` or `20260628`), a partial `shard_map` (i.e. `axis_names` is a strict subset of the mesh axes) whose `out_specs` references any of the newly-manual axes fails to lower on the GSPMD path (`jax_use_shardy_partitioner=False`) with an `AssertionError` in `named_sharding.py`.

This worked on `0.11.0.dev20260624` and earlier.

We understand the GSPMD path is being deprecated in favour of Shardy, but it is still present in the codebase and the fix is two lines (see below), so filing in case it's useful.

## Repro

```python
import os
os.environ["XLA_FLAGS"] = "--xla_force_host_platform_device_count=8"
os.environ["JAX_PLATFORMS"] = "cpu"

import jax
jax.config.update("jax_use_shardy_partitioner", False)

import jax.numpy as jnp
from jax.sharding import Mesh, PartitionSpec as P, NamedSharding
import numpy as np

mesh = Mesh(np.array(jax.devices()[:8]).reshape(2, 2, 2), ("a", "b", "c"))

@jax.jit
def f(x):
    # partial shard_map: {a,b} manual, c stays auto; out_spec references 'b'
    return jax.shard_map(
        lambda x: jax.lax.psum(x, "a"),
        mesh=mesh,
        in_specs=P("a", "b"),
        out_specs=P(None, "b"),
        check_vma=False,
        axis_names=frozenset({"a", "b"}),
    )(x)

x = jax.device_put(jnp.ones((4, 4)), NamedSharding(mesh, P("a", "b")))
print(f(x))  # AssertionError at named_sharding.py:427 on dev20260628
             # prints a (2,4) array of 2.0 on dev20260624
```

<details><summary>Traceback</summary>

```
  File ".../jax/_src/named_sharding.py", line 427, in named_sharding_to_xla_hlo_sharding
    assert {x[0] for x in replicated_mesh_axes}.issuperset(set(special_axes.keys()))
AssertionError
```
</details>

## Cause

In bbd2b2461, `_shard_map_lowering` was changed to reassign `ctx` so that `mlir.call_lowering` receives a `LoweringRuleContext` whose `module_context` has the inner `SPMDAxisContext`:

```python
sub_ctx = ctx.module_context.replace(axis_context=new_axis_context)
ctx = ctx.replace(module_context=sub_ctx)   # <— new in bbd2b2461
```

After `call_lowering`, `_xla_unshard(ctx, ...)` is called — but `ctx` now carries the **inner** `SPMDAxisContext` (whose `manual_axes` include `newly_manual_axes`), whereas previously it saw the outer context. `_make_scoped_manual_sharding` therefore marks the newly-manual axes as `AxisType.Manual`, and building a `NamedSharding` whose spec references one of those axes trips the assertion in `named_sharding_to_xla_hlo_sharding` (a manual axis must be replicated, not tiled in the spec).

The same reassignment also means `ctx.set_tokens_out(tokens_out)` now writes to the replaced object rather than the caller's `ctx`, which I believe would silently drop ordered-effect tokens for `shard_map` on the GSPMD path.

## Proposed fix

Use a fresh local instead of reassigning `ctx`, so `_xla_unshard` and `set_tokens_out` continue to see the outer context (matching `_xla_shard`, which is called before the reassignment):

```diff
--- a/jax/_src/shard_map.py
+++ b/jax/_src/shard_map.py
@@ -1102,10 +1102,10 @@ def _shard_map_lowering(ctx: mlir.LoweringRuleContext, *in_nodes,
   new_axis_context = sharding_impls.SPMDAxisContext(
       _get_spmdaxis_ctx_mesh(mesh), newly_manual_axes | set(mesh.manual_axes))
   sub_ctx = ctx.module_context.replace(axis_context=new_axis_context)
-  ctx = ctx.replace(module_context=sub_ctx)
+  inner_ctx = ctx.replace(module_context=sub_ctx)
   with _extend_axis_env(mesh, newly_manual_axes), config._check_vma(check_vma):
     out_nodes_, tokens_out = mlir.call_lowering(
-        "shmap_body", pe.close_jaxpr(jaxpr), None, ctx, sub_ctx, in_avals_,
+        "shmap_body", pe.close_jaxpr(jaxpr), None, inner_ctx, sub_ctx, in_avals_,
         out_avals_, ctx.tokens_in, *in_nodes_,
         dim_var_values=ctx.dim_var_values,
         const_lowering=ctx.const_lowering,
```

`call_lowering` reads `ctx.avals_out` (unchanged by the replace) and passes `ctx` through to `lower_with_sharding_in_types` → `wrap_with_sharding_op`. Those do read `ctx.module_context`, but never `.axis_context`, which is the only field that differs between `inner_ctx` and the original `ctx` — so the two are equivalent for `call_lowering`'s purposes. The `mod_ctx` parameter (still `sub_ctx`) is what carries the inner axis context into the called computation.

With this patch applied on top of `0.11.0.dev20260628`, the repro above lowers, compiles, runs, and produces the same result as `0.11.0.dev20260624`.

## Environment

- jax / jaxlib: `0.11.0.dev20260628` (fails), `0.11.0.dev20260624` (works)
- Platform: CPU (`--xla_force_host_platform_device_count=8`); no accelerator needed
- Python 3.13

**Comment by yashk2810** [comment]
Oops, sorry. It was my change. I am looking into similar failures internally so I'll fix it.

**Comment by yashk2810** [comment]
Also, you should switch to using shardy where this problem doesn't happen.
